### ISSUE TITLE
Converting to scram throws error due to a newline character in password

### ISSUE THREAD

**Issue by cdroege** [processone/ejabberd#996]
I am trying to convert a user database in mysql to the scram format. For this I have a database, that has the correct table coloumns for scram. I start the conversion via ejabberdctl debug:

```
ejabberd_auth_odbc:convert_to_scram(<<"localhost">>).
{aborted,{badarg,[{crypto,sha_mac_n,
                          [error,
                           << [REMOVED]
                             >>,
                           20],
                          []},
                  {scram,hi,3,[{file,"src/scram.erl"},{line,73}]},
                  {ejabberd_auth_odbc,password_to_scram,2,
                                      [{file,"src/ejabberd_auth_odbc.erl"},{line,396}]},
                  {ejabberd_auth_odbc,'-convert_to_scram/1-fun-0-',1,
                                      [{file,"src/ejabberd_auth_odbc.erl"},{line,450}]},
                  {lists,foreach,2,[{file,"lists.erl"},{line,1337}]},
                  {ejabberd_auth_odbc,'-convert_to_scram/1-fun-1-',0,
                                      [{file,"src/ejabberd_auth_odbc.erl"},{line,447}]},
                  {ejabberd_odbc,outer_transaction,3,
                                 [{file,"src/ejabberd_odbc.erl"},{line,445}]},
                  {ejabberd_odbc,run_sql_cmd,4,
                                 [{file,"src/ejabberd_odbc.erl"},{line,382}]}]}}
```

This error happens after it converted the first hundreds of users. It has a problem with passwords that contain a newline character ("\n"). There are a few dozen users with such a password.

Here is a test case with an empty database:

```
$ ejabberdctl register test localhost "bla
"
User test@localhost successfully registered

[add the missing table coloumns to the users table here]

$ echo 'select * from ejabberd.users where username="test";' | mysql -u root
username    password    created_at  serverkey   salt    iterationcount
test bla\n  2016-03-10 14:33:12         0

$ ejabberdctl debug
[...]
(ejabberd@localhost)1> ejabberd_auth_odbc:convert_to_scram(<<"localhost">>).
{aborted,{badarg,[{crypto,sha_mac_n,
                          [error,
                           <<224,198,212,158,107,84,199,123,219,64,207,103,61,94,82,
                             49,0,0,0,...>>,
                           20],
                          []},
                  {scram,hi,3,[{file,"src/scram.erl"},{line,73}]},
                  {ejabberd_auth_odbc,password_to_scram,2,
                                      [{file,"src/ejabberd_auth_odbc.erl"},{line,396}]},
                  {ejabberd_auth_odbc,'-convert_to_scram/1-fun-0-',1,
                                      [{file,"src/ejabberd_auth_odbc.erl"},{line,450}]},
                  {lists,foreach,2,[{file,"lists.erl"},{line,1337}]},
                  {ejabberd_auth_odbc,'-convert_to_scram/1-fun-1-',0,
                                      [{file,"src/ejabberd_auth_odbc.erl"},{line,447}]},
                  {ejabberd_odbc,outer_transaction,3,
                                 [{file,"src/ejabberd_odbc.erl"},{line,445}]},
                  {ejabberd_odbc,run_sql_cmd,4,
                                 [{file,"src/ejabberd_odbc.erl"},{line,382}]}]}}
```

I tested this with ejabberd 16.02 and 16.01.

**Comment by cdroege** [comment]
The "error" comes from jid:resourceprep here:

```
salted_password(Password, Salt, IterationCount) ->
    hi(jid:resourceprep(Password), Salt, IterationCount).
```

The password should be normalized with saslprep and not resourceprep. Saslprep seems to support newline characters, if I read the RFC correctly. The stringprep module does not support saslprep.

**Comment by cdroege** [comment]
Nevermind. Saslprep does not allow newline characters.

**Comment by badlop** [comment]
In summary, what's the result of your findings?
A) does XMPP allow newline characters in passwords (and then ejabberd has a problem)?
B) Or newline characters are not allowed (and then ejabberd is correct)?

**Comment by cdroege** [comment]
Both. It depends on the SASL mechanism.

From the [PLAIN](https://tools.ietf.org/html/rfc4616) RFC:
```
   The presented authentication identity and password strings, as well
   as the database authentication identity and password strings, are to
   be prepared before being used in the verification process.  The
   [SASLPrep] profile of the [StringPrep] algorithm is the RECOMMENDED
   preparation algorithm.  The SASLprep preparation algorithm is
   recommended to improve the likelihood that comparisons behave in an
   expected manner.  The SASLprep preparation algorithm is not mandatory
   so as to allow the server to employ other preparation algorithms
   (including none) when appropriate.  For instance, use of a different
   preparation algorithm may be necessary for the server to interoperate
   with an external system.
```

From the [SCRAM](https://tools.ietf.org/html/rfc5802) RFC:
```
[...]
     SaltedPassword  := Hi(Normalize(password), salt, i)
[...]
 o  Normalize(str): Apply the SASLprep profile [RFC4013] of the
      "stringprep" algorithm [RFC3454] as the normalization algorithm to
      a UTF-8 [RFC3629] encoded "str".  
```

SASLprep does not allow control characters such as newline characters.

Ejabberd does not use a preparation algorithm when using PLAIN (or when registering an account), but uses it for SCRAM. This causes a problem when converting the database and I found another problem, that I tested:

1. User registers an account with a newline character
2. Server saves this password as plaintext
3. Users client tries to use SCRAM authentication
4. Server responds with not-authorized, because the saved password does not pass the [jid:resourceprep(Password)](https://github.com/processone/ejabberd/blob/master/src/scram.erl#L39) check (resourceprep and SASLprep are very similiar, so this is OK)
5. PLAIN authentication by the clients works

There probably should always be a jid:resourceprep check, when a password is used, but this is a problem for existing password databases, because they already contain a newline character. But it should not be a problem adding this check to the registration of new users.

Note: I dont have much experience reading RFCs.

**Comment by zinid** [comment]
This issue is still requires some fixes, see #1295
